scenebuilder: init at 15.0.1 + scenic-view: init at 11.0.2

wirew0rm opened the pull request:
###### Motivation for this change
Add packages for [scenebuilder](https://gluonhq.com//products/scene-builder/) and [scenic-view](https://github.com/JonathanGiles/scenic-view/). Having a packaged version of scenebuilder is useful becaue it can be configured in eclipse, injellij Idea and other Java IDEs to design JavaFx UIs.
I also added scenic-view, because it is a very helpful tool for JavaFx UI design and also because the derivations are quite simillar so I think it is easier to review them together.

fixes: #90243
replaces: #51564

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

wirew0rm committed:
scenebuilder: init at 15.0.0

wirew0rm committed:
scenic-view: init at 11.0.2

wirew0rm committed:
maintainers: add wirew0rm

SuperSandro2000 commented on pkgs/development/tools/scenebuilder/default.nix:
```diff
@@ -0,0 +1,115 @@
+    name = "Scene Builder";
+    exec = "scenebuilder";
+    icon = "scenebuilder";
+    comment = "Scene Builder is a visual, drag'n'drop, layout tool for designing JavaFX application user     interfaces.";
```
```suggestion
    comment = "Scene Builder is a visual, drag'n'drop, layout tool for designing JavaFX application user interfaces.";
```

SuperSandro2000 commented on pkgs/development/tools/scenebuilder/default.nix:
```diff
@@ -0,0 +1,115 @@
+
+in stdenv.mkDerivation rec {
+  inherit pname version src;
+  name = "${pname}-${version}";
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/tools/scenicview/default.nix:
```diff
@@ -0,0 +1,112 @@
+  desktopItems = [ desktopItem ];
+
+  meta = with lib; {
+    description = "Scenic View is a JavaFX application designed to make it simple to understand the current state of your application scenegraph, and to also easily manipulate properties of the scenegraph without having to keep editing your code. This lets you find bugs, and get things pixel perfect without having to do the compile-check-compile dance.";
```
This is to long for description. Please use longDescription and short this line.

SuperSandro2000 commented on pkgs/development/tools/scenicview/default.nix:
```diff
@@ -0,0 +1,112 @@
+  meta = with lib; {
+    description = "Scenic View is a JavaFX application designed to make it simple to understand the current state of your application scenegraph, and to also easily manipulate properties of the scenegraph without having to keep editing your code. This lets you find bugs, and get things pixel perfect without having to do the compile-check-compile dance.";
+    homepage = "https://github.com/JonathanGiles/scenic-view/";
+    license = licenses.gpl3;
```
gpl3Only or gpl3Plus?

wirew0rm commented:
@SuperSandro2000 Thank you for your feedback, I hope I addressed all your points. I also renamed the scenicview derivation to scenic-view to be the same as the github repo and the package name and removed the gradle `-x tests` flag excluding tests which is not necessary for scenic-view, but scenebuilder won't build without it.

@FRidh @jonringer please excuse the erroneous review requests, i accidentally pushed a non related commit and it seems I cannot undo the request. Sorry for the noise.

SuperSandro2000 commented:
I am not that fluent in gradle or java in nixpkgs so I would appreciate it if someone else could look at this, too.

wirew0rm commented:
Simplified installPhase to use makeWrapper.

jonringer commented:
> @FRidh @jonringer please excuse the erroneous review requests, i accidentally pushed a non related commit and it seems I cannot undo the request. Sorry for the noise.

I'm used to it, don't worry

ShamrockLee commented:
The derivation is built successfully, but clicking `Open` from the menu causes `GLib-GIO-ERROR`.

Outputs:
```console
[nix-shell:~/.cache/nixpkgs-review/pr-110398]$ scenebuilder
Fontconfig warning: "/etc/fonts/fonts.conf", line 86: unknown element "blank"
Jan 26, 2021 11:01:16 PM com.oracle.javafx.scenebuilder.app.SceneBuilderApp logTimestamp
INFO: JavaFX Scene Builder started

(java:14910): GLib-GIO-ERROR **: 23:01:30.364: No GSettings schemas are installed on the system
Trace/breakpoint trap (core dumped)

```

wirew0rm commented:
Thanks for testing, it seems that due to javafx using gtk adding gappsWrapperArgs to the wrapper might be necessary. On my system this seems to be propagated somewhere in my system environment, so I cannot really tell if this is sufficient.

Just for reference, are you running on nixos or on a different distribution? I'm testing on nixos-unstable with i3 as a window manager (usually I'm the one with the missing desktop environment variables :D).

I only added the hook for scenebuilder, it might be also needed for scenic-view.

Another problem I've noticed is that font selection dialogs do not work, I get "Several fonts have same family and style". But I suspect that this is not really packaging related.

wirew0rm commented:
ok, i could reproduce your error by unsetting some environment variables:

``` bash
unset GIO_EXTRA_MODULES     
unset XDG_DATA_DIRS         
```
The former was not included in my inital fix, because i created the wrapper in the install phase, but the glib hook populates the wrapper aguments in the preFixup phase. 

I also tested scenic-view in that environment but it does not seem to require wrapping.

On a side note if I interpret [the 4. point in this nixpkgs manual paragraph](https://nixos.org/manual/nixpkgs/stable/#ssec-gnome-common-issues) libraries could be packaged in a way, that they do not require wrapping of packages which depend on them. So maybe this could also be fixed in the [openjfx derivation](https://github.com/NixOS/nixpkgs/blob/nixos-20.09/pkgs/development/compilers/openjdk/openjfx/14.nix) by hardcoding the gSettings path there, but I don't know enough about gtk and the openjfx build to know if that would be feasible.

ShamrockLee commented:
Thank for the fix. It now works alright.

The build result successfully reads the .fxml file I have written.

ShamrockLee commented on pkgs/development/tools/scenebuilder/default.nix:
```diff
@@ -0,0 +1,111 @@
+    homepage = "https://gluonhq.com/products/scene-builder/";
+    license = licenses.bsd3;
+    maintainers = with maintainers; [ wirew0rm ];
+  };
```
`meta.platforms` attribute should be addressed.
```suggestion
    platforms = platforms.all;
  };
```
> [`meta.platforms` is the] list of Nix platform types on which the package is supported. Hydra builds packages according to the platform specified. If no platform is specified, the package does not have prebuilt binaries.

https://nixos.org/manual/nixpkgs/stable/#chap-meta

As a source-build package, `lib.platforms.all` should be fine, unless it were known to break on certain platforms.

wirew0rm replied:
I was not sure if the jdks on other platforms also contain the javafx modules.

ShamrockLee replied:
Me neither. Since both (jdk for Linux and Mac) are in nixpkgs, it probably won't be a problem.
Extra dependencies such as openjfx can be added if it should fail on Mac.

ShamrockLee commented on pkgs/development/tools/scenebuilder/default.nix:
```diff
@@ -0,0 +1,112 @@
+{ lib, stdenv, fetchFromGitHub, jdk, jre, gradleGen, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper, glib, wrapGAppsHook }:
```
```suggestion
{ lib, stdenv, fetchFromGitHub, jdk, gradleGen, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper, glib, wrapGAppsHook }:
```

ShamrockLee commented on pkgs/development/tools/scenic-view/default.nix:
```diff
@@ -0,0 +1,105 @@
+{ lib, stdenv, fetchFromGitHub, jdk, jre, gradleGen, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper }:
```
```suggestion
{ lib, stdenv, fetchFromGitHub, jdk, gradleGen, makeDesktopItem, copyDesktopItems, perl, writeText, runtimeShell, makeWrapper }:
```

ShamrockLee reviewed (changes requested):
`jdk` should be enough. In addition, `jre` isn't used in both expressions.

(Sorry for not pointing them out in a single comment. It was until you mentioned jdk that I saw the duplication.)

wirew0rm commented:
Resolved confilcts and checked using [nixpkgs-hammering](https://github.com/jtojnar/nixpkgs-hammering). Resolved all issues, except for the custom phases, I think for gradle it is reasonable to use custom build an install phases.

SuperSandro2000 commented on r/scenebuilder:
```diff
@@ -0,0 +1 @@
+/nix/store/230ga7fzgvdfddd38bgpi4xgyv8w7y4w-scenebuilder-15.0.0
```
Does files are not supposed to be commited.

wirew0rm replied:
m( damn, fixed.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/need-help-with-packaging-scenebuilder/12289/9

SuperSandro2000 commented on pkgs/development/tools/scenic-view/default.nix:
```diff
@@ -0,0 +1,109 @@
+
+  meta = with lib; {
+    description = "JavaFx application to visualize and modify the scenegraph of running JavaFx applications.";
+    longDescription = "A JavaFX application designed to make it simple to understand the current state of your application scenegraph, and to also easily manipulate properties of the scenegraph without having to keep editing your code. This lets you find bugs, and get things pixel perfect without having to do the compile-check-compile dance.";
```
```suggestion
    longDescription = ''
      A JavaFX application designed to make it simple to understand the current state of your application scenegraph,
      and to also easily manipulate properties of the scenegraph without having to keep editing your code.
      This lets you find bugs, and get things pixel perfect without having to do the compile-check-compile dance.
    '';
```

SuperSandro2000 commented on pkgs/development/tools/scenebuilder/default.nix:
```diff
@@ -0,0 +1,117 @@
+  '';
+
+  postFixup = ''
+    echo "''${gappsWrapperArgs[@]}"
```
```suggestion
```

wirew0rm commented:
@SuperSandro2000 resolved your comments. I also realized that SceneBuilder 16.0.0 is already released and they changed the build system to maven. I've updated to the last gradle version, but I'll check if I can get the current version working.
